service/debugger/debugger.go

```diff
@@ -0,0 +1,453 @@
+// Debugger can be exposed by other services.
+type Debugger struct {
+	config       *Config
+	processFuncs chan func(*proctl.DebuggedProcess) error
+	stop         chan stopSignal
+}
+
+// Config provides the configuration to start a Debugger. If ProcessArgs are
+// specified, a new process is launched. If AttachPid is >0, the debugger will
+// attach to an existing process.
+type Config struct {
+	// ProcessArgs are the arguments to launch a new process.
+	ProcessArgs []string
+	// AttachPid is the pid of an existing process to which the debugger should
+	// attach.
+	AttachPid int
+}
+
+// stopSignal is used to stop the debugger.
+type stopSignal struct {
+	// KillProcess indicates whether to kill the debugee following detachment.
+	KillProcess bool
+}
+
+// New creates a new Debugger.
+func New(config *Config) *Debugger {
+	debugger := &Debugger{
+		processFuncs: make(chan func(*proctl.DebuggedProcess) error),
+		config:       config,
+		stop:         make(chan stopSignal),
+	}
+	return debugger
+}
+
+// withProcess facilitates thread-safe access to the DebuggedProcess. All
+// interaction with DebuggedProcess should occur via calls to withProcess, and
+// the functions placed on the processFuncs channel should be consumed and
+// executed from the same thread as the DebuggedProcess.
+//
+// This is convenient because it allows things like HTTP handlers in
+// goroutines to work with the DebuggedProcess with synchronous semantics.
+func (d *Debugger) withProcess(f func(*proctl.DebuggedProcess) error) error {
+	result := make(chan error)
+	d.processFuncs <- func(proc *proctl.DebuggedProcess) error {
+		result <- f(proc)
+		return nil
+	}
+	return <-result
+}
+
+// Run starts debugging a process until Detach is called.
+func (d *Debugger) Run() error {
+	// We must ensure here that we are running on the same thread during
+	// the execution of dbg. This is due to the fact that ptrace(2) expects
+	// all commands after PTRACE_ATTACH to come from the same thread.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// Create the process by either attaching or launching.
+	var process *proctl.DebuggedProcess
+	if d.config.AttachPid > 0 {
+		glog.Infof("attaching to pid %d", d.config.AttachPid)
+		p, err := proctl.Attach(d.config.AttachPid)
+		if err != nil {
+			return fmt.Errorf("couldn't attach to pid %d: %s", d.config.AttachPid, err)
+		}
+		process = p
+	} else {
+		glog.Infof("launching process with args: %v\n", d.config.ProcessArgs)
+		p, err := proctl.Launch(d.config.ProcessArgs)
+		if err != nil {
+			return fmt.Errorf("couldn't launch process: %s\n", err)
+		}
+		process = p
+	}
+
+	// Handle access to the process from the current thread.
+	for {
+		select {
+		case f := <-d.processFuncs:
+			// Execute the function
+			f(process)
+		case s := <-d.stop:
+			// Handle shutdown
+			glog.Info("debugger is stopping")
+
+			// Clear breakpoints
+			bps := []*proctl.BreakPoint{}
+			for _, bp := range process.BreakPoints {
```

Comment:
For what it's worth, breakpoints in this map will never be nil, only HWBreakPoints are potentially nil.